### PR TITLE
Add LogExpM1 transformation

### PR DESCRIPTION
**Pull request by junpenglao**
Softplus transformation (from non-negative to reals) might be more numerically stable than Log (see Fig. 9 in Kucukelbir et al. 2017).

### CODE REVIEW

**Comment by junpenglao** [comment]
Potentially also change the default transformation for Gamma(a, b) using softplus

**Comment by twiecki** [comment]
Can you link to the paper directly?

**Comment by junpenglao** [comment]
Yep - it is the ADVI paper actually: https://arxiv.org/pdf/1603.00788.pdf

**Comment by twiecki** [comment]
Makes sense. I think we should make this the default for Gamma then.

File: pymc3/distributions/transforms.py [hunk 1]
```diff
@@ -105,12 +105,27 @@ def jacobian_det(self, x):
 log = Log()
 
 
+class Softplus(ElemwiseTransform):
```
Comment (twiecki) [inline pymc3/distributions/transforms.py]: Not sure Softplus is the right name, above we name things according to the forward transform. `LogExp`, `SoftMinus`?

**Comment by springcoil** [comment]
Does this apply to anything else other than gamma?

On 27 Sep 2017 09:06, "Thomas Wiecki" <notifications@github.com> wrote:

Makes sense. I think we should make this the default for Gamma then.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/pymc-devs/pymc3/pull/2601#issuecomment-332443141>, or mute
the thread

**Comment by twiecki** [comment]
@springcoil Wondered the same thing. According to the paper any non-negative distribution with light tails.

**Comment by springcoil** [comment]
I think logexpsoftminus makes more sense as a name

**Comment by junpenglao** [comment]
How about `LogExpm1`?

**Comment by twiecki** [comment]
Want to change the Gamma too?

**Comment by junpenglao** [comment]
I asked @dustinvtran on twitter:
"Hey @dustinvtran, is Log(Exp(x)-1) the recommended transform for all Positive -> Real or it is just for Gamma as in Kucukelbir et al. 2017?"

I will either change the default for all positive distribution or just the Gamma.

**Comment by ferrine** [comment]
Nice question. That was a dilemma for me too when working on VI.

**Comment by dustinvtran** [comment]
I replied to @junpenglao saying, "I always recommend softplus as the default for [0, infty) -> R. See also a note I wrote here: https://groups.google.com/d/msg/stan-dev/6iP0DzW5Yec/ZkTzdJ_6AgAJ".

In the ADVI journal paper, we describe sensitivity of transformations with respect to their statistical implications. I recommend softplus in general not for these reasons but because it has good numerical underflow/overflow properties and has a well-behaving derivative. It's useful both for transforming random variables and for transforming trainable variables.

You might also be interested in Edward's [automated transformation notebook](https://github.com/blei-lab/edward/blob/master/notebooks/automated_transformations.ipynb). It (albeit briefly) mentions this.

**Comment by junpenglao** [comment]
Thanks a lot @dustinvtran!

**Comment by twiecki** [comment]
Even though well motivated, this change has pretty far-reaching implications. Not sure what we can do to make sure we're not surprising users. Definitely note it in the release-notes.

**Comment by junpenglao** [comment]
@twiecki otherwise we just leave the default as it is and put in the release note that this is now an option? I will wait for everybody's weight in before I fix the test.

**Comment by kyleabeauchamp** [comment]
Maybe add a list of distributions impacted by this under the hood to the release notes?

File: pymc3/distributions/transforms.py [hunk 2]
```diff
@@ -105,12 +105,31 @@ def jacobian_det(self, x):
 log = Log()
 
 
+class Logexpm1(ElemwiseTransform):
```
Comment (ColCarroll) [inline pymc3/distributions/transforms.py]: super nitpick, not very important, but easier to change now: `LogExpM1`, and `log_exp_m1`

Comment (junpenglao) [reply pymc3/distributions/transforms.py]: So the class name `LogExpM1` and `name = "log_exp_m1"`?

Comment (ColCarroll) [reply pymc3/distributions/transforms.py]: exactly -- convention is underscore to separate words in snake case, capital letter in CamelCase (again, this isn't important, but might give some clues to users figuring out how to parse the word :D)

**Comment by junpenglao** [comment]
This is ready.

**Review by ColCarroll — APPROVED** [review]

**Comment by ColCarroll** [comment]
Looks good -- can argue over default transformations after 3.2?

**Pull request merged**